### ISSUE TITLE
Statistics using longitude bins before averaging over distinct ASC regimes

### ISSUE THREAD

**Issue by fabiobdias** [willaguiar/ASC_and_heat_transport#29]
I'm trying to calculate the statistics using 5deg longitude bins rather than averaged over the whole regime as we've been doing. I've included below a plot of the monthly along-slope current for both the original and binned into 5-deg longitude.

![Screenshot 2024-02-07 at 9 52 19 am](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/dfe40040-7198-45aa-a7fb-d8b66e658c48)

![Screenshot 2024-02-07 at 9 52 30 am](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/57e0074e-0f1f-4434-b435-3c93ec7190d5)

The problem occurs when trying to calculate the linear regression statistics. There are 2 loops (borrowed from the original GMM_ASC.ipynb), which were originally looping the scipy.stats.linregress function through (a) ASC regimes (n=3) and (b) depth levels (n=50). 

For the binned CSHT/ASC, it means going through the whole binned contour (n=71). Using XXLargeMem, every stats.linregress takes ~14seconds to complete, which results in 14sec x 50 x 71 x 9 (doing it 9 times for _all, _annual, and _clima variables) = ~124 hours.

Tried to split these into 3 pbs scripts, and using hugemem, and it still ran out of walltime (max 48h).

I think one solution might be to vectorise the CHST/ASC inputs to the linregress, but when I tried I realised the CSHT has 75 vertical levels and ASC along-slope has only 50. Can anyone recall me why is that?

**Comment by fabiobdias** [comment]
update: I got these statistics done (I wasn't ".load()" after doing the binning which was delaying the calculations - thanks Will!). 

So, here we have done the statistics (r2 and linear regression slope) for each of the 5-deg bins and then averaged out for each ASC regime (using the time-mean mask). The high correlations in the upper-100m for the surface regime decreased substantially, but the high correlation at ~800m depth remains.

![layer_wise_CSHT_vs_U_corr_5degBins](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/febb7fba-8717-43f0-8a92-acf9d7655600)

**Comment by adele-morrison** [comment]
Interesting that there's not much difference between the regimes or timescales here. I wonder why it's different from [Ellie's analysis](https://github.com/willaguiar/ASC_and_heat_transport/issues/22#issuecomment-1931357407) where the annual average correlations seemed to improve going to regional analysis? Is it dependent on the longitude bin size we choose?

It would also be interesting to show the regional data on a map (or plotted against longitude on x-axis). Say, the r^2 value averaged between 800-1000m for each 5deg longitude bin, vs longitude?

**Comment by fabiobdias** [comment]
I was also wondering about the similarity among regimes/timescales (thought I was doing smth wrong with the masks but it seems alright). I will try smaller and bigger bins to test its dependence.

Here is what the r^2 averaged below 760m for the 5deg longitude bins looks like:

<img width="895" alt="Screenshot 2024-02-10 at 11 00 17 am" src="https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/39fd3b0f-317a-4383-b27c-46b22d14e789">

**Comment by fabiobdias** [comment]
I was also wondering about the similarity among regimes/timescales (thought I was doing smth wrong with the masks but it seems alright). I will try smaller and bigger bins to test its dependence.

Here is what the r^2 averaged below 760m for the 5deg longitude bins looks like:

<img width="955" alt="Screenshot 2024-02-10 at 11 08 02 am" src="https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/47f0ad2b-dc13-486e-9947-5a31c08081ad">

**Comment by ongqingyee** [comment]
Maybe larger longitudinal bins? The Totten region I selected is ~13deg longitude wide, and the rest are 30 deg. Esp as the CSHT was already binned by 3 degs by my understanding...?

**Comment by fabiobdias** [comment]
Here is the attempt with 15-deg longitudinal bins:

1) first look at the binned current along-slope:

![Screenshot 2024-02-12 at 10 56 16 am](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/cc7fed32-db81-4ed9-91ee-10b6e0f61fbb)

2) here's the regional r^2:
![Screenshot 2024-02-12 at 10 57 38 am](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/f85b7e2e-c201-4b1a-aff1-b22ce78140b0)

3) finally all statics per depth/regime/timescales:
![Screenshot 2024-02-12 at 10 58 22 am](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/ba0c79eb-bcc4-4698-9b9a-eb2b48215a47)

Some correlations do get better, such as near-surface in both surface-intensified and deep regimes (monthly climatology) and at depth (reverse & deep regimes, all timescales).

**Comment by fabiobdias** [comment]
and here is another attempt using 2-deg longitudinal bins:

![Screenshot 2024-02-12 at 1 07 49 pm](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/f0bc5dbe-f836-4eda-b6db-4d0492a8c311)

![Screenshot 2024-02-12 at 1 08 10 pm](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/814531a2-f75e-4eb3-8f08-83d710227998)

![Screenshot 2024-02-12 at 1 08 54 pm](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/84f12c13-704b-4e69-9066-9409a9b332f7)

**Comment by taimoorsohail** [comment]
I don't think we've done the analysis yet (correct me if I'm wrong) of assessing the correlations for the original resolution of longitudes using daily data, then averaging over the regimes. I tried something similar but it was for monthly data, and with time-varying, not time-mean regime masks. It would be good to see what the correlation looks like for the original longitude grid (i.e., without any coarsening), with daily data and time-mean regime masks, as it would provide a good baseline for comparison with these binned longitude diagnostics, but also with the correlations provided by will [here](https://github.com/willaguiar/ASC_and_heat_transport/issues/20#issuecomment-1925931481).

**Comment by fabiobdias** [comment]
Thanks for the suggestion @taimoorsohail - I will try to use daily data and get these correlations in the original resolution. I also had a chat with @willaguiar today and I've applied the regimes masks after binning in longitudes (and then calculated the correlations); I will test masking it before binning to see if that isn't affecting the results.

**Comment by fabiobdias** [comment]
I've re-calculated these statistics, now applying the regimes masks u-along/CSHT) before the longitudinal binning. The good news is that some of the high r^2 present in more than 1 regime at the same longitudinal bin (as some of the r^2 maps shown above) are avoided now. However, overall the results haven't changed that much, with a general increase of the r^2 as we use larger bins. Plots are shown in the comment below. I will try to repeat those using the daily data next...

**Comment by fabiobdias** [comment]
for 2deg bins:

![image](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/0101d3cf-642e-40fe-b6bf-bc13fd13e221)

![image](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/c5389845-2eeb-4613-ab83-02bdbee0cf09)

**Comment by fabiobdias** [comment]
for 5-deg bins:

![image](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/9610a45e-e41e-4c10-8685-e7fc8f27e0c7)

![image](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/f461b18f-27fc-4b9c-bb52-3bb8824bd24b)

**Comment by fabiobdias** [comment]
for 15-deg bins:

![image](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/f9c1ddb4-b041-4953-872e-bb9214dab49f)

![image](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/a3f4a104-c5cb-4f4c-8931-e188b08ffbaf)

**Comment by fabiobdias** [comment]
Figure comparing statistics over different longitudinal bin sizes - shows increase of the r^2 at depth with larger bins:

![layer_wise_CSHT_vs_U_corr_BinsComparison_v2](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/a3819aa0-5f40-41ce-ae13-3d7347590fa1)

**Comment by fabiobdias** [comment]
yeah I was thinking that... should decrease at some point, right? I will try some more bin sizes...

**Comment by taimoorsohail** [comment]
Cool! At least for these bin widths, if we take r^2=0.5 as our “significance threshold” then it only adds a few more depth levels in the reverse and deep regimes for 15-deg bins (with the exception of Deep/Monthly Climatology which does seem to include a lot more depth levels). Would be cool to see what happens in the extremes as Adele mentioned.

**Comment by ongqingyee** [comment]
Below are the circumpolar maps for 800-1400m depth-averaged correlations, variance of ASC and variance of CSHT using data binned longitudinally (thanks Fabio!). I calculate the variance at each depth level before taking the depth-average. The binned quantities are interpolated to cut off at the edge of each longitudinal bin so that I could plot them nicely. The colors of the outer ring show the time mean regime mask using GMM. I don't know if there a pattern in the correlation w.r.t the regimes, but the variance line up with the local processes at least.

5 deg bins
![Pasted image 20240322132533](https://github.com/willaguiar/ASC_and_heat_transport/assets/54887782/5ec6a5f2-7d8b-44c9-9cc6-428258b5a1e3)

10 deg bins
![image](https://github.com/willaguiar/ASC_and_heat_transport/assets/54887782/161b0393-9d2e-452b-8e84-0201316994e7)

15 deg bins
![Pasted image 20240322140713](https://github.com/willaguiar/ASC_and_heat_transport/assets/54887782/38f96594-1bc5-4117-965d-3777ab102120)

**Comment by fabiobdias** [comment]
I've tried a couple more longitudinal bin sizes (5, 10, 15, 20, 30, 40 degrees). Different bins are represent with slightly different of red (r^2) and green (regression slope) tones:

![layer_wise_CSHT_vs_U_corr_BinsComparison_v2_new](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/43af0448-b6e4-4e43-a38d-a435857706e6)

The relatively high r^2 below 700m does converge towards larger longitudinal bins at all timescales for the surface and reverse regimes. But it doesn't seem to be converging for the deep regime at seasonal time (monthly clim).

These results aren't yet weighted by the number of points in each regime (as we discussed last meeting), which I'm doing now and should post another update soon.

**Comment by fabiobdias** [comment]
it seems like yes, the longitude values along the contour are indeed uniformly spaced, e.g.:

`CSHT_months.lon`
`xarray.DataArray 'lon' lon: 1428`
`array([-278.5 , -278.25, -278.  , ...,   77.75,   78.  ,   78.25])`

 which in this case wouldn't change the results above if we weight-average the statistics.

**Comment by adele-morrison** [comment]
Yes the heat transport is binned into evenly spaced longitude values when including the zonal convergence term.

But within each bin, not all of the longitude points will be the same regime. E.g. if there’s 20 longitude values in a 5deg bin, then maybe only 2 of those longitudes correspond to the warm regime, so we wouldn’t want to weight this bin the same as another bin that has 20 warm point regime longitude values in it.

**Comment by fabiobdias** [comment]
ahh got it! thanks for the clarification @adele-morrison

**Comment by fabiobdias** [comment]
Ok results are quite sensitive if we weight-averaged the statistics. In the reverse and deep regimes, increasing the bin size causes r^2's to decrease (so smaller bins, i.e. 5deg, has larger correlations). The surface regime seems to be less sensitive to the weighting. R^2 in general are smaller than before, although max correlation at depth still present. Maximum correlation (r^2 ~ 0.7) is found at reverse regime at depths > 800m at seasonal timescale:

![layer_wise_CSHT_vs_U_corr_BinsComparison_v2_wavg](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/a4a81b6d-96fa-4f4c-bf54-9b77535924a0)

**Comment by adele-morrison** [comment]
With the new bin weighting, the correlations seem MUCH lower than previously (roughly 0.2-0.4 less!). If we increase the bin size to 360deg, then we should expect these to converge on our old larger correlations before we started binning right? Does that mean there's a bug somewhere, or should we keep increasing bin sizes to see when the correlation increases again?

**Comment by fabiobdias** [comment]
Just posting here the fixed the weight-average calculation, which now applying the weights correctly, and then sum up for each regime and divide by the sum of the weights. Correlations now are similar to what we had before (without the weighting ):

![layer_wise_CSHT_vs_U_corr_BinsComparison_v2_wavg](https://github.com/willaguiar/ASC_and_heat_transport/assets/8631257/f333a3f3-e0ba-4fea-a349-e503ddee0e8b)

During the hackathon today we agreed to shows a version of the Figure 2 with different bin sizes in the suppl. material, and use 20deg bins for the main figures.

**Comment by willaguiar** [comment]
Here is an updated version, with the daily deseasoned data on the bottom, in the 20 deg bin config. PDFs on the way....

<img width="930" alt="Screenshot 2024-04-22 at 4 16 24 PM" src="https://github.com/willaguiar/ASC_and_heat_transport/assets/70033934/d80fae95-3a58-4a6c-b0bf-ae8087be88d6">

**Comment by adele-morrison** [comment]
Curious that the seasonal correlations for the surface regime are so weak now, because they were ~0.8 before. I guess that's due to the change from zonal avg to 20deg bins?

**Comment by adele-morrison** [comment]
Out of curiosity, what does the daily look like with the new binning without the seasonality removed?

**Comment by willaguiar** [comment]
> Curious that the seasonal correlations for the surface regime are so weak now, because they were ~0.8 before. I guess that's due to the change from zonal avg to 20deg bins?

yeah, it looks like by comparing with [this fig](https://github.com/willaguiar/ASC_and_heat_transport/issues/20#issuecomment-1998872056). I wonder why the binning changes so much the surface result ( cause even 5deg binning decreases the correlation compared to the original one).


> Out of curiosity, what does the daily look like with the new binning without the seasonality removed?

Below is the daily 5 to 40 deg binning with (top) and without seasonality(bottom):

<img width="930" alt="Screenshot 2024-04-23 at 2 32 53 PM" src="https://github.com/willaguiar/ASC_and_heat_transport/assets/70033934/9e19a70a-3d51-43d5-be83-1da9cbe52110">
<img width="969" alt="Screenshot 2024-04-23 at 2 32 27 PM" src="https://github.com/willaguiar/ASC_and_heat_transport/assets/70033934/ab057927-4fd2-4db6-b580-3c4a4e6b7baa">

For me it looks like the deseasoning is doing very little for the correlations ( can only see small changes in the upper 200m of the surface, and upper 400m of the deep)


Additional info: for deseasoning I calculated the monthly climatology, interpolated it to daily, and then removed it from the daily data. 

Below are the plots for the deseasoned CSHT and U ( vertical sum and mean).
<img width="966" alt="Screenshot 2024-04-23 at 2 55 44 PM" src="https://github.com/willaguiar/ASC_and_heat_transport/assets/70033934/63b82b64-959c-4014-b350-36b700a7e183">
 

<img width="986" alt="Screenshot 2024-04-23 at 2 55 34 PM" src="https://github.com/willaguiar/ASC_and_heat_transport/assets/70033934/91242ebc-260a-4794-b1dd-b84fa449245d">
